bundles/org.openhab.binding.enphase/src/main/java/org/openhab/binding/enphase/internal/discovery/EnphaseDevicesDiscoveryService.java

```diff
@@ -0,0 +1,137 @@
+        }
+    }
+
+    @Override
+    public @Nullable ThingHandler getThingHandler() {
+        return envoyHandler;
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    @Override
+    protected void startScan() {
+        removeOlderResults(getTimestampOfLastScan());
+        final EnvoyBridgeHandler envoyHandler = this.envoyHandler;
+
+        if (envoyHandler == null || !envoyHandler.isOnline()) {
+            logger.debug("Envoy handler not available or online: {}", envoyHandler);
+            return;
+        }
+        final ThingUID uid = envoyHandler.getThing().getUID();
+
+        scanForInverterThings(envoyHandler, uid);
+        scanForDeviceThings(envoyHandler, uid);
+    }
+
+    private void scanForInverterThings(final EnvoyBridgeHandler envoyHandler, final ThingUID bridgeID) {
+        final Map<String, @Nullable InverterDTO> inverters = envoyHandler.getInvertersData(true);
+
+        if (inverters == null) {
+            logger.debug("No inverter data for Enphase inverters in discovery for Envoy {}.", bridgeID);
+        } else {
+            for (final Entry<String, @Nullable InverterDTO> entry : inverters.entrySet()) {
+                discover(bridgeID, entry.getKey(), THING_TYPE_ENPHASE_INVERTER, "Inverter ");
+            }
+        }
+    }
+
+    /**
+     * Scans for other device things ('other' as in: no inverters).
+     *
+     * @param envoyHandler
+     * @param bridgeID
+     */
+    private void scanForDeviceThings(final EnvoyBridgeHandler envoyHandler, final ThingUID bridgeID) {
+        final Map<String, @Nullable DeviceDTO> devices = envoyHandler.getDevices(true);
+
+        if (devices == null) {
+            logger.debug("No device data for Enphase devices in discovery for Envoy {}.", bridgeID);
+        } else {
+            for (final Entry<String, @Nullable DeviceDTO> entry : devices.entrySet()) {
+                final DeviceDTO dto = entry.getValue();
+                final EnphaseDeviceType type = dto == null ? null : EnphaseDeviceType.safeValueOf(dto.type);
+
+                if (type == EnphaseDeviceType.NSRB) {
+                    discover(bridgeID, entry.getKey(), THING_TYPE_ENPHASE_RELAY, "Relay ");
+                }
+            }
+        }
+    }
+
+    private void discover(final ThingUID bridgeID, final String serialNumber, final ThingTypeUID typeUID,
+            final String label) {
+        final String shortSerialNumber = defaultPassword(serialNumber);
+        final ThingUID thingUID = new ThingUID(typeUID, bridgeID, shortSerialNumber);
+        final Map<String, Object> properties = new HashMap<>(1);
+
+        properties.put(CONFIG_SERIAL_NUMBER, serialNumber);
+        final DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withBridge(bridgeID)
+                .withRepresentationProperty(CONFIG_SERIAL_NUMBER).withProperties(properties)
+                .withLabel("Enphase " + label + shortSerialNumber).build();
```

Comment:
Since you are using a withRepresentationProperty, the shortSerialNumber will already get displayed in the discovery results in OH3 directly under the label... Does it need to be displayed twice making the label longer?

Reply:
At some places only the labels are visible. Here it's a the label for an individual solar panel. So if one has 14 solar panels it would show 14 same labels. Therefor I added part of the serial number to the label.